File: MCPForUnity/Editor/Tools/ManageMultiplayer.cs
using System;
using System.Collections.Generic;
using MCPForUnity.Editor.Helpers;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// Tool for managing Unity Netcode for GameObjects and multiplayer features.
    /// </summary>
    [McpForUnityTool("manage_multiplayer")]
    public static class ManageMultiplayer
    {
        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
                return new ErrorResponse("Parameters cannot be null.");

            string action = ParamCoercion.CoerceString(@params["action"], null)?.ToLowerInvariant();
            if (string.IsNullOrEmpty(action))
                return new ErrorResponse("'action' parameter is required.");

            try
            {
                return action switch
                {
                    "setup_network_manager" => SetupNetworkManager(@params),
                    "add_network_object" => AddNetworkObject(@params),
                    "add_network_transform" => AddNetworkTransform(@params),
                    "add_network_rigidbody" => AddNetworkRigidbody(@params),
                    "add_network_animator" => AddNetworkAnimator(@params),
                    "add_network_variable" => AddNetworkVariable(@params),
                    "create_rpc" => CreateRpc(@params),
                    "create_lobby" => CreateLobby(@params),
                    "relay_setup" => RelaySetup(@params),
                    "get_connected_clients" => GetConnectedClients(@params),
                    _ => new ErrorResponse($"Unknown action: '{action}'.")
                };
            }
            catch (Exception e)
            {
                McpLog.Error($"[ManageMultiplayer] Action '{action}' failed: {e}");
                return new ErrorResponse($"Error: {e.Message}");
            }
        }

        private static object SetupNetworkManager(JObject @params)
        {
            // Check for existing NetworkManager
            var existingManager = GameObject.FindObjectOfType<MonoBehaviour>();
            
            var networkManagerGO = new GameObject("NetworkManager");
            
            return new SuccessResponse("NetworkManager setup.", new
            {
                gameObject = "NetworkManager",
                requiredPackages = new[]
                {
                    "com.unity.netcode.gameobjects",
                    "com.unity.transport"
                },
                components = new[]
                {
                    "NetworkManager",
                    "UnityTransport"
                },
                instruction = "Add NetworkManager and UnityTransport components manually or via Package Manager"
            });
        }

        private static object AddNetworkObject(JObject @params)
        {
            string targetPath = ParamCoercion.CoerceString(@params["targetObject"], "");
            
            if (string.IsNullOrEmpty(targetPath))
                return new ErrorResponse("'targetObject' is required.");

            var target = GameObject.Find(targetPath);
            if (target == null)
                return new ErrorResponse($"GameObject not found: {targetPath}");

            return new SuccessResponse($"NetworkObject component info for '{targetPath}'.", new
            {
                targetObject = targetPath,
                component = "NetworkObject",
                instruction = "Add Unity.Netcode.NetworkObject component to the GameObject",
                codeExample = "gameObject.AddComponent<NetworkObject>();"
            });
        }

        private static object AddNetworkTransform(JObject @params)
        {
            string targetPath = ParamCoercion.CoerceString(@params["targetObject"], "");
            bool syncX = ParamCoercion.CoerceBool(@params["syncPositionX"], true);
            bool syncY = ParamCoercion.CoerceBool(@params["syncPositionY"], true);
            bool syncZ = ParamCoercion.CoerceBool(@params["syncPositionZ"], true);
            bool syncRot = ParamCoercion.CoerceBool(@params["syncRotation"], true);
            bool syncScale = ParamCoercion.CoerceBool(@params["syncScale"], false);
            bool interpolate = ParamCoercion.CoerceBool(@params["interpolate"], true);

            return new SuccessResponse("NetworkTransform configured.", new
            {
                targetObject = targetPath,
                syncPosition = new { x = syncX, y = syncY, z = syncZ },
                syncRotation = syncRot,
                syncScale = syncScale,
                interpolate = interpolate,
                component = "NetworkTransform"
            });
        }

        private static object AddNetworkRigidbody(JObject @params)
        {
            string targetPath = ParamCoercion.CoerceString(@params["targetObject"], "");

            return new SuccessResponse("NetworkRigidbody configured.", new
            {
                targetObject = targetPath,
                component = "NetworkRigidbody",
                note = "Syncs Rigidbody physics across network"
            });
        }

        private static object AddNetworkAnimator(JObject @params)
        {
            string targetPath = ParamCoercion.CoerceString(@params["targetObject"], "");

            return new SuccessResponse("NetworkAnimator configured.", new
            {
                targetObject = targetPath,
                component = "NetworkAnimator",
                note = "Syncs Animator parameters across network"
            });
        }

        private static object AddNetworkVariable(JObject @params)
        {
            string varName = ParamCoercion.CoerceString(@params["variableName"], "networkVar");
            string varType = ParamCoercion.CoerceString(@params["variableType"], "int");
            string writePermission = ParamCoercion.CoerceString(@params["writePermission"], "Server");

            string netVarType = varType switch
            {
                "int" => "NetworkVariable<int>",
                "float" => "NetworkVariable<float>",
                "bool" => "NetworkVariable<bool>",
                "string" => "NetworkVariable<FixedString64Bytes>",
                "Vector3" => "NetworkVariable<Vector3>",
                "Quaternion" => "NetworkVariable<Quaternion>",
                _ => $"NetworkVariable<{varType}>"
            };

            return new SuccessResponse("NetworkVariable configured.", new
            {
                variableName = varName,
                variableType = netVarType,
                writePermission = writePermission,
                codeExample = $"public {netVarType} {varName} = new {netVarType}(default, NetworkVariableReadPermission.Everyone, NetworkVariableWritePermission.{writePermission});"
            });
        }

        private static object CreateRpc(JObject @params)
        {
            string rpcName = ParamCoercion.CoerceString(@params["rpcName"], "MyRpc");
            string rpcType = ParamCoercion.CoerceString(@params["rpcType"], "ServerRpc");
            var rpcParamsArray = @params["rpcParams"] as JArray;

            var paramList = new List<string>();
            if (rpcParamsArray != null)
            {
                foreach (var param in rpcParamsArray)
                {
                    string name = param["name"]?.ToString() ?? "param";
                    string type = param["type"]?.ToString() ?? "int";
                    paramList.Add($"{type} {name}");
                }
            }

            string paramsString = string.Join(", ", paramList);
            string attribute = rpcType == "ServerRpc" ? "[ServerRpc]" : "[ClientRpc]";
            string suffix = rpcType == "ServerRpc" ? "ServerRpc" : "ClientRpc";

            return new SuccessResponse($"{rpcType} created.", new
            {
                rpcName = rpcName + suffix,
                rpcType = rpcType,
                codeExample = $@"{attribute}
public void {rpcName}{suffix}({paramsString})
{{
    // RPC implementation
}}"
            });
        }

        private static object CreateLobby(JObject @params)
        {
            string lobbyName = ParamCoercion.CoerceString(@params["lobbyName"], "My Lobby");
            int maxPlayers = ParamCoercion.CoerceInt(@params["maxPlayers"], 4);
            bool isPrivate = ParamCoercion.CoerceBool(@params["isPrivate"], false);

            return new SuccessResponse("Lobby configuration.", new
            {
                lobbyName = lobbyName,
                maxPlayers = maxPlayers,
                isPrivate = isPrivate,
                requiredPackage = "com.unity.services.lobby",
                codeExample = $@"var lobby = await LobbyService.Instance.CreateLobbyAsync(""{lobbyName}"", {maxPlayers}, new CreateLobbyOptions {{ IsPrivate = {isPrivate.ToString().ToLower()} }});"
            });
        }

        private static object RelaySetup(JObject @params)
        {
            return new SuccessResponse("Unity Relay setup info.", new
            {
                requiredPackage = "com.unity.services.relay",
                steps = new[]
                {
                    "1. Enable Relay in Unity Dashboard",
                    "2. Initialize Unity Services",
                    "3. Create allocation with RelayService.Instance.CreateAllocationAsync()",
                    "4. Get join code with RelayService.Instance.GetJoinCodeAsync()",
                    "5. Configure transport with allocation data"
                }
            });
        }

        private static object GetConnectedClients(JObject @params)
        {
            return new SuccessResponse("Connected clients info.", new
            {
                codeExample = "NetworkManager.Singleton.ConnectedClientsIds",
                note = "Access at runtime through NetworkManager"
            });
        }
    }
}

File: Server/src/services/tools/manage_multiplayer.py
"""
manage_multiplayer.py - Multiplayer & Networking Tool

Handles:
- Unity Netcode for GameObjects
- Network objects
- RPCs and Client/Server communication
- Lobby and matchmaking
"""

from typing import Any
from mcp.types import TextContent
from .utils import generate_tool_response


def get_tool_definition():
    """Returns the tool definition for manage_multiplayer."""
    return {
        "name": "manage_multiplayer",
        "description": """Manage Unity Netcode for GameObjects, networking, and multiplayer features.

Actions:
- setup_network_manager: Create and configure NetworkManager
- add_network_object: Add NetworkObject component to GameObject
- add_network_transform: Add NetworkTransform for sync
- add_network_rigidbody: Add NetworkRigidbody for physics sync
- add_network_animator: Add NetworkAnimator for animation sync
- add_network_variable: Add NetworkVariable to script
- create_rpc: Create ServerRpc or ClientRpc method
- create_lobby: Create multiplayer lobby
- join_lobby: Join existing lobby
- leave_lobby: Leave current lobby
- list_lobbies: List available lobbies
- start_host: Start as host
- start_server: Start as dedicated server
- start_client: Start as client
- stop_network: Stop networking
- get_connected_clients: Get list of connected clients
- kick_client: Kick a client from server
- spawn_networked: Spawn networked object
- despawn_networked: Despawn networked object
- relay_setup: Setup Unity Relay for NAT traversal""",
        "inputSchema": {
            "type": "object",
            "properties": {
                "action": {
                    "type": "string",
                    "description": "Action to perform",
                    "enum": [
                        "setup_network_manager",
                        "add_network_object",
                        "add_network_transform",
                        "add_network_rigidbody",
                        "add_network_animator",
                        "add_network_variable",
                        "create_rpc",
                        "create_lobby",
                        "join_lobby",
                        "leave_lobby",
                        "list_lobbies",
                        "start_host",
                        "start_server",
                        "start_client",
                        "stop_network",
                        "get_connected_clients",
                        "kick_client",
                        "spawn_networked",
                        "despawn_networked",
                        "relay_setup"
                    ]
                },
                "targetObject": {
                    "type": "string",
                    "description": "Target GameObject path"
                },
                "syncPositionX": {"type": "boolean", "description": "Sync X position"},
                "syncPositionY": {"type": "boolean", "description": "Sync Y position"},
                "syncPositionZ": {"type": "boolean", "description": "Sync Z position"},
                "syncRotation": {"type": "boolean", "description": "Sync rotation"},
                "syncScale": {"type": "boolean", "description": "Sync scale"},
                "interpolate": {"type": "boolean", "description": "Enable interpolation"},
                "variableName": {
                    "type": "string",
                    "description": "Network variable name"
                },
                "variableType": {
                    "type": "string",
                    "description": "Network variable type",
                    "enum": ["int", "float", "bool", "string", "Vector3", "Quaternion"]
                },
                "writePermission": {
                    "type": "string",
                    "description": "Write permission for variable",
                    "enum": ["Server", "Owner"]
                },
                "rpcName": {
                    "type": "string",
                    "description": "RPC method name"
                },
                "rpcType": {
                    "type": "string",
                    "description": "RPC type",
                    "enum": ["ServerRpc", "ClientRpc"]
                },
                "rpcParams": {
                    "type": "array",
                    "description": "RPC parameters",
                    "items": {
                        "type": "object",
                        "properties": {
                            "name": {"type": "string"},
                            "type": {"type": "string"}
                        }
                    }
                },
                "lobbyName": {
                    "type": "string",
                    "description": "Lobby name"
                },
                "lobbyId": {
                    "type": "string",
                    "description": "Lobby ID"
                },
                "maxPlayers": {
                    "type": "integer",
                    "description": "Maximum players in lobby"
                },
                "isPrivate": {
                    "type": "boolean",
                    "description": "Is lobby private"
                },
                "lobbyData": {
                    "type": "object",
                    "description": "Custom lobby data"
                },
                "serverAddress": {
                    "type": "string",
                    "description": "Server IP address"
                },
                "serverPort": {
                    "type": "integer",
                    "description": "Server port"
                },
                "clientId": {
                    "type": "integer",
                    "description": "Client ID"
                },
                "prefabPath": {
                    "type": "string",
                    "description": "Prefab path for spawning"
                },
                "spawnPosition": {
                    "type": "object",
                    "description": "Spawn position",
                    "properties": {
                        "x": {"type": "number"},
                        "y": {"type": "number"},
                        "z": {"type": "number"}
                    }
                },
                "transportType": {
                    "type": "string",
                    "description": "Network transport type",
                    "enum": ["UnityTransport", "WebSocket"]
                }
            },
            "required": ["action"]
        }
    }


async def handle_tool_call(arguments: dict[str, Any]) -> list[TextContent]:
    """Handle manage_multiplayer tool calls."""
    action = arguments.get("action")
    
    if not action:
        return generate_tool_response({
            "success": False,
            "error": "'action' parameter is required"
        })

    params = {"action": action}
    
    param_keys = [
        "targetObject", "syncPositionX", "syncPositionY", "syncPositionZ",
        "syncRotation", "syncScale", "interpolate", "variableName", "variableType",
        "writePermission", "rpcName", "rpcType", "rpcParams", "lobbyName",
        "lobbyId", "maxPlayers", "isPrivate", "lobbyData", "serverAddress",
        "serverPort", "clientId", "prefabPath", "spawnPosition", "transportType"
    ]
    
    for key in param_keys:
        if key in arguments and arguments[key] is not None:
            params[key] = arguments[key]

    return generate_tool_response({
        "tool": "manage_multiplayer",
        "params": params
    })
